compiler/rustc_hir_typeck/src/mem_categorization.rs

```diff
@@ -131,9 +131,15 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         match ty {
             Some(ty) => {
                 let ty = self.resolve_vars_if_possible(ty);
-                if ty.references_error() || ty.is_ty_var() {
+                if ty.references_error() {
                     debug!("resolve_type_vars_or_error: error from {:?}", ty);
                     Err(())
+                } else if ty.is_ty_var() {
+                    debug!("resolve_type_vars_or_error: infer var from {:?}", ty);
+                    self.tcx()
+                        .dcx()
+                        .span_delayed_bug(self.tcx().hir().span(id), "encountered type variable");
+                    Err(())
```

Comment:
do we ever return `Err(())` in places which should successfully compile? If not, please change `McResult` to `Result<T, ErrorGuaranteed>`